### ISSUE TITLE
Distributed Multi-machine Training with PyTorch Lightning?

### ISSUE THREAD

**Issue by andrewjong** [microsoft/pai#4463]
**Short summary about the issue/question**:
PyTorch Lightning claims to make it easy to run multi-node training with PyTorch. [Here are its docs](https://pytorch-lightning.readthedocs.io/en/latest/multi_gpu.html#distributed-modes) on the subject. Any advice on how I could make this work with OpenPAI?

I looked at the distributed TensorFlow examples, where ps_hosts and worker_hosts are used, but those seems specific to that TensorFlow example.

**Comment by abuccts** [comment]
You could refer to this [PyTorch Horovod example](https://github.com/microsoft/pai/blob/master/marketplace-v2/horovod-pytorch-synthetic-benchmark.yaml).

**Comment by scarlett2018** [comment]
> **Short summary about the issue/question**:
> PyTorch Lightning claims to make it easy to run multi-node training with PyTorch. [Here are its docs](https://pytorch-lightning.readthedocs.io/en/latest/multi_gpu.html#distributed-modes) on the subject. Any advice on how I could make this work with OpenPAI?
> 
> I looked at the distributed TensorFlow examples, where ps_hosts and worker_hosts are used, but those seems specific to that TensorFlow example.

It is highly appreciated if you'd love to share this with all the OpenPAI users when it works (after refer the example provided by @abuccts ) =)!.

**Comment by andrewjong** [comment]
Thanks for the suggestions. I'm looking into it. Can OpenPAI run horovod with the `horovodrun` command? E.g. `horovodrun -np 8 -H hostname1:4,hostname2:4 python train.py`? Or does it have to be run with `mpirun`?

**Comment by andrewjong** [comment]
@abuccts interestingly I could NOT successfully run the PyTorch Horovod Example. The example hangs and does not complete successfully. Here is stdout:

```
[INFO] USER COMMAND START

Model: resnet50
Batch size: 64
Number of GPUs: 8
Running warmup...

pai15:187:393 [0] misc/ibvwrap.cu:63 NCCL WARN Failed to open libibverbs.so[.1]
pai15:187:393 [0] NCCL INFO Using internal Network Socket
pai15:187:393 [0] NCCL INFO NET/Socket : 0 interfaces found

pai15:187:393 [0] transport/net_socket.cu:38 NCCL WARN NET/Socket : no interface found
pai15:187:393 [0] NCCL INFO transport/net_socket.cu:114 -> 3
pai15:187:393 [0] NCCL INFO bootstrap.cu:18 -> 3
pai15:187:393 [0] NCCL INFO bootstrap.cu:133 -> 3
pai15:187:393 [0] NCCL INFO bootstrap.cu:154 -> 3
```

I had to use horovodrun instead for the command. This command completed successfully:
```
horovodrun -np 8 -H master-0:4,worker-0:4 python pytorch_synthetic_benchmark.py \
--model <% $parameters.model %> --batch-size <% $parameters.batchsize %>
```

I am using OpenPAI 0.14.0.

**Comment by abuccts** [comment]
> Can OpenPAI run horovod with the `horovodrun` command?

You can run any command as it is valid, `horovodrun` is just a wrapper for `mpirun`

> I could NOT successfully run the provided PyTorch Horovod Example. The example hangs and does not complete successfully

According to your log `transport/net_socket.cu:38 NCCL WARN NET/Socket : no interface found`, you need to change the network interface [parameter](https://github.com/microsoft/pai/blob/master/marketplace-v2/horovod-pytorch-synthetic-benchmark.yaml#L16) to fit your ethernet environment if it's not `eth0`.

`horovodrun` is a simplified `mpirun`, it works in simple case but doesn't work if you need to tune parameters in `mpirun`. pls refer to [horovodrun docs](https://horovod.readthedocs.io/en/latest/mpirun.html)

**Comment by andrewjong** [comment]
Update: I got PyTorch Lightning working with OpenPAI by using Horovod as the backend! Here's my [.yaml file](https://github.com/andrewjong/PT-Lightning-for-OpenPAI/blob/master/pytorch_lightning_horovod.yaml), working on OpenPAI 0.14.0.

**Comment by abuccts** [comment]
> Update: I got PyTorch Lightning working with OpenPAI by using Horovod as the backend! Here's my [.yaml file](https://github.com/andrewjong/PT-Lightning-for-OpenPAI/blob/master/pytorch_lightning_horovod.yaml), working on OpenPAI 0.14.0.

seems you duplicate four tasks, you can write one task role named `server` with 4 instances, and change host list from `-H server1-0:4,server2-0:4,server3-0:4,server4-0:4` to `-H server-0:4,server-1:4,server-2:4,server-3:4`.

**Comment by andrewjong** [comment]
Great point, thanks! That's much cleaner.

**Comment by andrewjong** [comment]
Besides the difference in brevity, is there any other advantage of using instances over tasks? Such as runtime/latency?

**Comment by abuccts** [comment]
> Besides the difference in brevity, is there any other advantage of using instances over tasks? Such as runtime/latency?

just in case if you want to run on 100 nodes so you only need to change one number
using multiple tasks by setting instances may also save bandwidth in large scale situation when querying status after submission, comparing to multiple task roles
